Install YARN on cgcloud-jenkins/toil_jenkins_slave (resolves #149, #161)

fnothaft opened the pull request:
Resolves #149, #161 

Installs Apache Hadoop 2.6.2 on the toil_jenkins_slave image by pulling down the prebuilt artifacts from an Apache mirror. YARN is a subcomponent of Hadoop. This does not start the YARN nodemanager/resourcemanager, as they are started downstream in the toil test suites.

Also see https://github.com/BD2KGenomics/toil/issues/796.

Still WIP. A quick question: I need to set the `HADOOP_HOME` environment variable. Is the best way to do this to cat to `~/.profile`? I know that I can use the fabric context manager utilities to set an envar when using fabric, but I need this environment variable to be set for the YARN tests in toil to run. Otherwise, I also need to modify a Hadoop config file; I will add that soon.

fnothaft committed:
Install YARN on cgcloud-jenkins/toil_jenkins_slave (resolves #149)

Installs Apache Hadoop 2.6.2 on the toil_jenkins_slave image by pulling down
the prebuilt artifacts from an Apache mirror. YARN is a subcomponent of Hadoop.
This does not start the YARN nodemanager/resourcemanager, as they are started
downstream in the toil test suites.

Also see https://github.com/BD2KGenomics/toil/issues/796.

fnothaft commented on jenkins/src/cgcloud/jenkins/toil_jenkins_slave.py:
```diff
@@ -48,7 +48,8 @@ def _post_install_packages( self ):
         self.__install_parasol( )
         self.__patch_distutils( )
         self.__configure_gridengine( )
-
```
Note to self: clean up spurious whitespace diff.

hannes-ucsc commented:
> Is the best way to do this to cat to ~/.profile

https://github.com/BD2KGenomics/cgcloud/search?utf8=%E2%9C%93&q=%22%2Fetc%2Fenvironment%22

hannes-ucsc commented:
`echo … >> /etc/environment` will do in a pinch but my __patch_etc_environment method is idempotent. You may want to pull that up in the class hierarchy and use that instead.

hannes-ucsc commented on jenkins/src/cgcloud/jenkins/toil_jenkins_slave.py:
```diff
@@ -72,6 +73,14 @@ def __install_parasol( self ):
 
+    @fabric_task
+    def __install_yarn ( self ):
+        yarn_version = '2.6.2'
```
Should probably be global.

hannes-ucsc commented on jenkins/src/cgcloud/jenkins/toil_jenkins_slave.py:
```diff
@@ -72,6 +73,14 @@ def __install_parasol( self ):
+    def __install_yarn ( self ):
+        yarn_version = '2.6.2'
+        run( 'mkdir /opt/apache-hadoop/' )
+        run( 'curl http://apache.osuosl.org/hadoop/common/hadoop-%s/hadoop-%s.tar.gz'
```
There is some Apache mirror lookup functionality in SparkBox which might be useful for this. Will need to be pulled up, so it's available here, too. Or put it in an ApacheSoftwareBox mix-in.

fnothaft commented:
Thanks @hannes-ucsc for the pointer to the '__patch_etc_environment'. I will pull that up and address your other comments!

jpdna commented:
@fnothaft - were you going to address these issues above before I do some further testing of toil/YARN, or did you want me to look into the above?  I'm suspect I don't know enough cgcloud-jenkins and it will take more time to answer questions then me just reviewing the changes you make, but I will dig into it if you want.  

Honestly, I'm not certain what is meant above by "pulling up" then `__patch_etc_environment` function the class hierarchy - does this mean to add spark_box as another parent to ToilJenksSlave class when it is declared at:  https://github.com/fnothaft/cgcloud/blob/6913f75bd0bca344c8ab7a9e459fd039d4801926/jenkins/src/cgcloud/jenkins/toil_jenkins_slave.py

Also - not sure how I'd go about testing Jenkins locally here - if you want me to work on this can you point me to what I'd need to setup to test?

hannes-ucsc commented:
If I may chime in, the "pull up" refactoring is explained [here](https://en.wikipedia.org/wiki/Pull_Up_refactoring).

You won't change the inheritance hierarchy, just remove the method from one class and move it into a class higher up in the hierarchy that is a common ancestor for the ToilJenkinsSlave class and the class the method is currently defined in. That way it will be accessible in both.

fnothaft committed:
Install Spark on Toil-Jenkins box (resolves #161)

Installs Spark on the Toil Jenkins box by downloading from an Apache mirror
for Spark and adds PySpark to the PYTHONPATH. Uses Spark 1.6.1/Hadoop 2.6.

fnothaft commented:
@hannes-ucsc ping for review. This resolves #161 as well now—I've updated the title. I think this is good to go from my side.

hannes-ucsc commented:
Looks like a missing import:

http://jenkins.cgcloud.info/job/cgcloud/296/console

hannes-ucsc commented on core/src/cgcloud/core/apache.py:
```diff
@@ -0,0 +1,60 @@
+
+        Some mirrors may be down or serve crap, so we may need to retry this a couple of times.
+        """
+        assert install_path is not None
```
Variable doesn't exist

hannes-ucsc commented on core/src/cgcloud/core/apache.py:
```diff
@@ -0,0 +1,60 @@
+        return url
+
+    def __apache_s3_mirror_url( self, path ):
+        file_name = os.path.basename( path )
```
os needs to be imported

hannes-ucsc commented on core/src/cgcloud/core/generic_boxes.py:
```diff
@@ -286,3 +286,44 @@ def __fix_stupid_locale_problem( self ):
+class GenericEnvironmentPatcherBox( Box ):
+
+    @classmethod
+    def __patch_etc_environment( cls, env_file, dirs=None, env_pairs=None, envar=None ):
```
Method should  be protected, not private.

hannes-ucsc commented on core/src/cgcloud/core/generic_boxes.py:
```diff
@@ -286,3 +286,44 @@ def __fix_stupid_locale_problem( self ):
+        r"""
+        >>> f=StringIO('FOO = " BAR " \n  PATH =foo:bar\nBLA="FASEL"')
+        >>> f.seek(0,2) # seek to end as if file was opened with 'a'
+        >>> SparkBox._SparkBox__patch_etc_environment( f, [ "new" ] )
```
Bad reference. But see above, when method becomes protected, this name mangling hack is not needed anymore.

hannes-ucsc commented on core/src/cgcloud/core/generic_boxes.py:
```diff
@@ -286,3 +286,44 @@ def __fix_stupid_locale_problem( self ):
         sudo( 'localedef -c -i en_US -f UTF-8 en_US.UTF-8' )
+
+
+class GenericEnvironmentPatcherBox( Box ):
```
Let's stuff this into RcLocalBox. I have a grand plan for that class to become LSBBox (as in Linux Software Base standard). Then you won't have to mix it in because RcLocalBox is already included in the hierarchy.

hannes-ucsc commented on spark/src/cgcloud/spark/spark_box.py:
```diff
@@ -33,7 +34,7 @@
 hdfs_replication = 1
 
-hadoop_version = '2.6.0'
+hadoop_version = '2.6.2'
```
Ahh, you snuck that in. I guess I'm ok with a minor bump like that.

fnothaft replied:
Ah, sorry! Didn't mean to sneak such in. But, I would prefer to bump to Hadoop 2.6.2 as that's what we're running yonder here at Berkeley.

fnothaft replied:
Er, sorry, this is indeed unimportant. Let me unroll this.

fnothaft commented:
Just pushed updates. Ping @hannes-ucsc for review.

hannes-ucsc commented:
Still numerous typos. I'll take it from here. I highly, highly recommend the use of an Python IDE like PyCharm, or in lieu of that, actually running the code that you write or change ;-). But I shouldn't call the kettle black. At least we're even now.

fnothaft commented:
Yeah, sorry about that. I know I'm lobbing a pile of... uhm... bad code over the wall at you.

hannes-ucsc commented:
Continued in #177
